```{r echo=FALSE}
library(knitr)
opts_chunk$set(echo=FALSE)
```

---
title: "IL-2 Influences of the Transcriptional Activity in Human MSC"
author: "Lu Wang"
date: "Monday, July 21, 2014"
output: html_document
---
## Experimental Design

![Experimental Design][img01]

## Libraries and Raw Data

In this analysis, we started with `.bam` files containing RNA-seq reads aligned to hg19. The first metadata file is the `libs` file which contains the name and total mapped reads of all libraries.

```{r cache=TRUE}
libs <- read.delim("./kazak_data/tmap_lib_size.tsv", header=TRUE,sep="\t")
```

```{r colors}
library(RColorBrewer)
c6 <- brewer.pal(6, "Paired")
c6 <- c6[c(2, 1, 6, 5, 4, 3)]

pal1 <- colorRampPalette(c6[1:2])
pal2 <- colorRampPalette(c6[3:4])
pal3 <- colorRampPalette(c6[5:6])

c15 <- c(pal1(7)[c(1:2, 5:7)], pal2(6)[c(1:2, 5:6)], pal3(8)[c(1:3, 6:8)])
```

In this data set, we have **`r nrow(libs)`** sequencing runs that correspond to **six** different RNA libraries/conditions:

1. SR
2. SR with IL-2 treatment
3. SEN
4. SEN with IL-2 treatment
5. REJ
6. REJ with IL-2 treatment

Each file contains mapped reads were pooled together, compared against RefSeq gene model(`RefSeq.bed` file downloaded from UCSC genome browser) and generated a counts matrix using `bedtools`. 

## Distribution of Raw Counts

The raw counts for each transcript is calculated by couting the number of reads overlapped with the **exons** of both protein-coding and non-coding **RefSeq transcripts**.

```{r readCountsMatrix, cache=TRUE}
tb <- read.table("./kazak_data/counts_matrix.txt",header=FALSE)
counts <- as.matrix(tb[,2:16])
rownames(counts) <- tb[,1]
rm(tb)
colnames(counts) <- rownames(libs)
counts <- counts[order(rownames(counts)),]
head(counts, n=2)
```


```{r CountsDensity}
library(reshape)
library(ggplot2)

plotDistribution <- function(data, title = ""){
    par(mar=c(8.1,6.1,4.1,2.1))
    par(las=2)
    boxplot(log2(data.frame(data)), col = c15, names = (colnames(data)), main = paste(title))
    data.df <- data.frame(id = rownames(data), data, stringsAsFactors = FALSE)
    mdf <- melt(data.df, id = "id", variable_name = "condition")

    ggplot(mdf, aes(log2(value))) +
        geom_density(aes(color = condition)) + 
        title(main = paste(title), xlab = "") + 
        scale_colour_manual(values = c15) +
        theme_bw() +
        theme(panel.grid.major = element_blank())
}

plotDistribution(counts+1, "Distribution of Read Counts")
```

In this analysis, **technical replicates for each condition were combined** to increase the total signal. We started with **`r nrow(counts)`** RefSeq transcript annotations.

```{r filterLibs, fig.width = 6}
c15 <- c6
sr_tr = grep("sr_treated", colnames(counts))
sr_un = grep("sr_untreated", colnames(counts))
sen_tr = grep("sen_treated", colnames(counts))
sen_un = grep("sen_untreated", colnames(counts))
rej_tr = grep("rej_treated", colnames(counts))
rej_un = grep("rej_untreated", colnames(counts))


counts <- data.frame( sr_treated = apply(counts[, sr_tr], 1, sum),
                      sr_untreated = apply(counts[, sr_un], 1, sum),
                      sen_treated = apply(counts[, sen_tr], 1, sum),
                      sen_untreated = apply(counts[, sen_un], 1, sum),
                      rej_treated = apply(counts[, rej_tr], 1, sum),
                      rej_untreated = apply(counts[, rej_un], 1, sum))
head(counts, 2)
plotDistribution(counts+1, "Distribution of Read Counts")
```

## ERCC Counts

The ERCC counts for each library were downloaded from the **Ion Proton** server with the help from James. The distribution of ERCC counts is shown below:

```{r readERCC, fig.width = 6}
ercc <- read.table("./kazak_data/ercc_counts.txt", header = TRUE)

ercc <- data.frame( sr_treated = apply(ercc[, sr_tr], 1, sum),
                    sr_untreated = apply(ercc[, sr_un], 1, sum),
                    sen_treated = apply(ercc[, sen_tr], 1, sum),
                    sen_untreated = apply(ercc[, sen_un], 1, sum),
                    rej_treated = apply(ercc[, rej_tr], 1, sum),
                    rej_untreated = apply(ercc[, rej_un], 1, sum))
head(ercc, 2)

plotDistribution(ercc+1, "Distribution of ERCC Counts")
```

In general, the distribution of **`r nrow(ercc)`**ERCC counts follows the trend of total read counts.

## Determining Lower Limit of Detection

The true counts of the ERCC sequences were obtaind from the `ERCC_Controls_Analysis.txt` file downloaded from [manufacturer's website][001]

```{r downloadERCC, cache = TRUE}
con = "http://tools.lifetechnologies.com/downloads/ERCC_Controls_Analysis.txt"
true.ercc <- read.delim(con, header = TRUE)
```

The scatterplots below compared the observed ERCC counts versus the theoretical number of molecules from the manufacturer's documentation.

```{r reformatERCC}
true.ercc <- true.ercc[, c(2, 4)]
true.ercc <- true.ercc[order(true.ercc$ERCC.ID), ]

head(true.ercc)
colnames(true.ercc) = c("ercc.id", "mix1.attmoles")
```


```{r obsERCCvsTrue, fig.height = 6, fig.width = 9}
library(reshape)
library(lattice)

low.cut <- 3
lib.cut <- 2

ercc.df <- cbind(ercc.id = rownames(ercc), ercc, stringsAsFactors = FALSE)

erccOvsE <- function(data, title = ""){
    temp <- merge(data, true.ercc, by.x = colnames(data)[grep("id", colnames(data))], by.y = "ercc.id")
    
    mdf <- melt(temp[, c(1, grep("sr|sen|rej", colnames(temp)))], id = colnames(data)[grep("id", colnames(data))], variable_name = "sample")
    
    mdf <- cbind(mdf, expected = rep(temp$mix1.attmoles, ncol(ercc)))
    
    mdf$sample <- ordered(mdf$sample)
    
    xyplot( log2(value + 1) ~ log2(expected+1) | sample, data = mdf, 
            layout = c(3,2), 
            as.table = TRUE, 
            pch = 20, 
            col = rgb(70, 130, 180, 100, maxColorValue = 255), 
            par.strip.text = list(cex = 0.7), 
            xlim = c(0, 15),
            ylim = c(0, 15),
            xlab = expression(log[2]*"Expected Counts"), 
            ylab = expression(log[2]*"Observed Counts"), 
            main = paste(title), 
            panel = function(x, y, ...) {
                panel.xyplot(x, y, ...)
                panel.abline(h = log2(low.cut+1), lty = 2)
                panel.lmline(x, y, col.line = rgb(0,0,0,0.2))
    })
}
erccOvsE(ercc.df, "ERCC Dose Response")
```

The **dashed line** in panels above shows the empirical cutoff for low read count sequences. The ERCC sequences with more than **`r low.cut`** counts in each library demonstrated linear relationship between the expected and observed read counts.

```{r plotPairwiseCorr, fig.height = 9, fig.width = 12}
plotPairs <- function(data, title = ""){
    pairs(log2(data), col = rgb(0,0.5,1,0.2), pch = 16, main = paste(title))
}
```


# Normalization

```{r exonLengths, cache=TRUE}
exon.lengths <- t(read.delim("./kazak_data/exon_lengths.tsv",header=TRUE,sep="\t"))
```

The exon length of each RefSeq transcripts were calculated using the `RefSeq.gtf` file downloaded from UCSC genome browser. 

```{r getGeneLengths}
counts.df <- data.frame(id = rownames(counts), counts, stringsAsFactors = FALSE)
lengths.df <- data.frame(id = rownames(exon.lengths), exon.lengths)
mdf <- merge(counts.df, lengths.df, by.x = "id", by.y = "id")
counts.df <- mdf
head(counts.df, 2)
```

For ERCC Spike-Ins, the lengths of the ERCC sequences were obtaind from the `ERCC92.gtf` file downloaded from [manufacturer's website][001]

```{r getERCClength}
tb <- read.table("http://tools.invitrogen.com/downloads/ERCC92.gtf", header = FALSE)
tb <- tb[, c(1, 5)]
colnames(tb) <- c("id", "length")
mdf <- merge(ercc.df, tb, by.x = "ercc.id", by.y = "id" )
ercc.df <- mdf
head(ercc.df, 2)
```


```{r RefSeq2HGNC_symbol, cache = TRUE}
library(biomaRt)


pcg <- grep("NM_", counts.df$id)
nc <- grep("NR_", counts.df$id)

pcg.id <- counts.df$id[pcg]
nc.id <- counts.df$id[nc]

getSymbol <- function(ids, id.type, attributes){
    ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
    res <- getBM( attributes = c(paste(id.type), paste(attributes)), filters = id.type, values = ids, mart = ensembl )
}


res1 <- getSymbol(pcg.id, "refseq_mrna", attributes = "hgnc_symbol")
res2 <- getSymbol(nc.id, "refseq_ncrna", attributes = "hgnc_symbol")
```

```{r cleanHGNC_symbol}
cleanBiomartRes <- function(data, id.type, attribute){
    # Remove nchar=0 empty results
    data[, attribute] <- ifelse(nchar(data[, paste(attribute)]) == 0, NA, data[, attribute])
    # Sort and remove duplicated id
    data <- data[order(data[, id.type], data[, attribute]), ]
    data <- data[!duplicated(data[, id.type]), ]
}
res1 <- cleanBiomartRes(res1, "refseq_mrna", "hgnc_symbol")
res2 <- cleanBiomartRes(res2, "refseq_ncrna", "hgnc_symbol")
```

```{r assembleHGNC_symbol}
#Assemble results for mRNA
pcg <- merge(counts.df[pcg, ], res1, by.x = "id", by.y = "refseq_mrna", all.x = TRUE)

fillMissingSymbol <- function(data){
    data$hgnc_symbol <- ifelse(is.na(data$hgnc_symbol), data$id, data$hgnc_symbol)
    return(data)
}
pcg <- fillMissingSymbol(pcg)
rm(res1)

#Assemble results for ncRNA
nc <- merge(counts.df[nc, ], res2, by.x = "id", by.y = "refseq_ncrna", all.x = TRUE)
nc <- fillMissingSymbol(nc)
rm(res2)
#Assemble mRNA and ncRNA
counts.df <- rbind(pcg, nc)
rm(pcg)
rm(nc)

ercc.df <- data.frame(ercc.df, hgnc_symbol = ercc.df$ercc.id, stringsAsFactors = FALSE)
colnames(ercc.df) <- gsub("ercc.id", "id", colnames(ercc.df))
colnames(counts.df) <- gsub("Length", "length", colnames(counts.df))
counts.df <- rbind(ercc.df, counts.df)

dataCols <- grep("sr|sen|rej", colnames(counts.df), perl = TRUE)
annoCols <- grep("id|symbol|length", colnames(counts.df), perl = TRUE)

groups <- colnames(counts.df)[dataCols]
groups <- gsub("_[0-9]+", "", groups)
groups <- factor(groups, levels = unique(groups))
```

```{r message = FALSE}
library(edgeR)
dge <- DGEList(counts = counts.df[, dataCols], group = groups, genes = counts.df[, annoCols])
```

## Filtering Low Expression Sequences


```{r FilterCounts, fig.width = 6}
#Filter Low Count Seqeunces
#keep <- apply(dge$counts, 1, min) > 3
keep <- rowSums(cpm(dge) > 1) >= 1
dge <- dge[keep, ]

plotDistribution(dge$counts, "Distribution of Filtered Counts")

counts.df <- data.frame(dge$counts, dge$genes)
```

After filtering, **`r sum(keep)`** transcripts were kept for downstream analysis.

## Expression of Genes
Since different RefSeq transcripts for the same gene symbol shares most of their reads, therefore, among the **`r nrow(counts.df)`** transcripts, the transcript with highest overall count for each gene symbol were used for that particular gene symbol. 

```{r getGeneExpression}
o <- order(rowSums(dge$counts), decreasing = TRUE)
dge <- dge[o, ]
dge <- dge[! duplicated(dge$genes$hgnc_symbol), ]

dge$samples$lib.size <- colSums(dge$counts)
```


## RPKM Normalization

The **RPKM** calculation for each RefSeq transcript was calculated using the method from *Mortazavi et. al*. 

For the *i* th transcript in library *j*, the RPKM is calculated as

$RPKM_{i,j} = \frac{exonReadsCount_{i,j}}{librarySize_{j} \times exonLength_{i} \times 10^{-9}}$


All RPKM values for transcripts were stored in the matrix below:

```{r getRPKM}
rpkms <- edgeR::rpkm(dge, gene.length = dge$genes$length, normalized.lib.sizes = FALSE)
rownames(rpkms) <- dge$genes$id
```


```{r RpkmDensity, fig.width = 6}
library(reshape)
library(ggplot2)

plotDistribution(rpkms, "Distribution of RPKM")
plotDistribution(rpkms[grep("ERCC-", dge$genes$id), ], "RPKM Normalized ERCC Sequences")
```

## TMM Normalization


```{r TMM}
dge <- calcNormFactors(dge)
dge$samples
tmm.multiplier = 1000000/dge$sample$lib.size/dge$sample$norm.factors
names(tmm.multiplier) = rownames(dge$samples)
tmm.multiplier
tmm <- cpm(dge, normalized.lib.sizes = TRUE)
```

```{r TMMDensity, fig.width = 6}
library(reshape)
library(ggplot2)

plotDistribution(tmm, "TMM Normalized Gene Expression")

plotDistribution(tmm[grep("ERCC-", dge$genes$id), ], "TMM Normalized ERCC Sequences")

tmm.df <- data.frame(tmm, dge$genes)
```

```{r erccTMMvsExpected, fig.height = 6, fig.width = 9}
erccOvsE(tmm.df, "ERCC Expected vs. TMM Normalized")
```

## DESeq Scaling Factor Normalization

```{r DESeq, message = FALSE}
library(DESeq)
temp <- dge$counts
rownames(temp) <- dge$genes$hgnc_symbol
cds <- newCountDataSet(temp, conditions = groups)

cds <- estimateSizeFactors(cds)
sizeFactors(cds)
deseq <- counts(cds, normalized = TRUE)
```

```{r DESeqDensity, fig.width = 6}
library(reshape)
library(ggplot2)

plotDistribution(deseq, "DESeq Normalized Counts")
plotDistribution(deseq[grep("ERCC-", rownames(deseq)), ], "DESeq Normalized ERCC")

deseq.df <- data.frame(deseq, dge$genes)
```


## LOESS Normalization

```{r LOESS, message = FALSE}
library(affy)
loess <- normalize.loess(dge$counts+0.5, subset = grep("ERCC-", dge$genes$id), maxit = 3, verbose = FALSE)
```

```{r LOESSDensity, fig.width = 6}
library(reshape)
library(ggplot2)

plotDistribution(loess, "LOESS Normalized Gene Expression")
plotDistribution(loess[grep("ERCC-", dge$genes$id), ], "LOESS Normalized ERCC Sequences")

loess.df <- data.frame(loess, dge$genes )
```

```{r erccLOESSvsExpected, fig.height = 6, fig.width = 9}
erccOvsE(tmm.df, "ERCC Expected vs. LOESS Normalized")
```

## RPKM + LOESS Normalization

```{r rpkmLOESS, message = FALSE}
library(affy)
rpkms.loess <- normalize.loess(rpkms+0.001, subset = grep("ERCC-", rownames(rpkms)), maxit = 3, verbose = FALSE)
```

```{r rpkmLOESSDensity, fig.width = 6}
library(reshape)
library(ggplot2)

plotDistribution(rpkms.loess, "RPKM + LOESS Normalized Gene Expression")
plotDistribution(rpkms.loess[grep("ERCC-", rownames(rpkms.loess)), ], "RPKM+ LOESS Normalized ERCC")

rpkms.loess.df <- data.frame(rpkms.loess, dge$genes)
```

# Exploratory Analysis
## Variance estimation

```{r estVar}
cdsBlind = estimateDispersions(cds, method = "blind", sharingMode="fit-only")
plotDispEsts(cdsBlind)
```
<!--

## Variance Stablizing & Clustering

```{r varStablize}
#cdsBlind = estimateDispersions(cds, method = "blind")
vsd = varianceStabilizingTransformation(cdsBlind)
```

```{r fig.height = 4, message = FALSE}
library(vsn)
par(mfrow=(c(1, 2)))
notAllZero = (rowSums(counts(cds)) > 0)
meanSdPlot(log2(counts(cds)[notAllZero, ] + 1))
meanSdPlot(vsd[notAllZero, ])
```



## Pairwise distance between samples
```{r distSamples, message = FALSE, fig.height = 6}
library(gplots)
hmcol = colorRampPalette(c6[1:3])(100)
dists = dist(t(exprs(vsd)))
mat = as.matrix(dists)
rownames(mat) = colnames(mat) = pData(cds)$condition
heatmap.2(mat, trace = "none", col = rev(hmcol), margin = c(13, 13))
```
-->
```{r fig.height = 7}
#plotPCA(vsd)
```


```{r PCAallGene, fig.width = 5, fig.height = 3}
pca1 <- prcomp(mat, scale = TRUE)
df <- data.frame( x = pca1$rotation[, 1], y = pca1$rotation[, 2], condition = factor(rownames(pData(cds)), levels = rownames(pData(cds))))
g <- ggplot(df, aes(x, y))
g <- g + geom_point(aes(color = condition),
                    size = 4, 
                    alpha = 4/5 )
g <- g + scale_color_manual( values = c15)
g <- g + theme_bw() + theme(panel.grid.major = element_blank()) 
g + labs(main = "", x = "PC1", y = "PC2")

#plot(pca1$rotation[, 1], pca1$rotation[, 2], col = c15, pch = 16, xlab = "PC1", ylab = "PC2")
#legend("topright", legend = rownames(pData(cds)), col = c15, pch = 16)
```
<!--

## Top 300 highly expressed genes

```{r highGenes, fig.height = 10}
library(gplots)
select = order(rowMeans(counts(cds)), decreasing = TRUE)[1:300]
heatmap.2(exprs(vsd)[select, ], col = hmcol, trace = "none", margin = c(10, 6))
```

## Gene Clusters of Top 300 Genes

```{r geneCluster, fig.height = 10}
mat <- exprs(vsd[select, ])
zmat <- (mat - rowMeans(mat))/apply(mat, 1, sd)
heatmap.2(zmat, col = hmcol, trace = "none", margin = c(10, 6))
```
-->

# Differential Expression Analysis

## Calling Differential Expression with DESeq

### SR_treated vs. SR_untreated

```{r DEanalysis, cache = TRUE}

res = nbinomTest( cdsBlind, "sr_untreated", "sr_treated")
```

```{r plotDESeqRes}
plotMA(res)
hist(res$padj, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)

getDEset <- function(res, pval.cut){
    res.sig <- res[res$padj < pval.cut, ]
    up <- res.sig[res.sig$log2FoldChange > 0, ]
    dn <- res.sig[res.sig$log2FoldChange < 0, ]
    mdf.up <- merge(deseq.df, up, by.x="hgnc_symbol", by.y="id")
    mdf.dn <- merge(deseq.df, dn, by.x="hgnc_symbol", by.y="id")
    mdf.up <- mdf.up[order(-mdf.up$foldChange, -mdf.up$baseMean), ]
    mdf.dn <- mdf.dn[order(mdf.dn$foldChange, -mdf.dn$baseMean), ]
    res <- list(mdf.up, mdf.dn)
}

res <- getDEset(res, 0.05)
up.sr.des <- res[[1]]
dn.sr.des <- res[[2]]

head(up.sr.des)
head(dn.sr.des)
```

In total, there are **`r nrow(up.sr.des)+nrow(dn.sr.des)`** statistically significant differentially expressed genes between **SR_treated and SR_untreated**. Among these genes, there are **`r nrow(up.sr.des)`** up-regulated and **`r nrow(dn.sr.des)`**.

### SEN_treated vs. SEN_untreated

```{r DEsen, cache = TRUE}

res = nbinomTest( cdsBlind, "sen_untreated", "sen_treated")
```

```{r plotDEsen}
plotMA(res)
hist(res$padj, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)

res <- getDEset(res, 0.05)

up.sen.des <- res[[1]]
dn.sen.des <- res[[2]]

head(up.sen.des)
head(dn.sen.des)
```

In total, there are **`r nrow(up.sen.des)+nrow(dn.sen.des)`** statistically significant differentially expressed genes between **SEN_treated and SEN_untreated**. Among these genes, there are **`r nrow(up.sen.des)`** up-regulated and **`r nrow(dn.sen.des)`** upon treatment.

### SEN_untreated vs. SR_untreated

```{r DEsenSR, cache = TRUE}

res = nbinomTest( cdsBlind, "sr_untreated", "sen_untreated")
```

```{r plotDEsenSR}
plotMA(res)
hist(res$padj, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)

res <- getDEset(res, 0.05)

up.untr.des <- res[[1]]
dn.untr.des <- res[[2]]

head(up.untr.des)
head(dn.untr.des)
```

In total, there are **`r nrow(up.untr.des)+nrow(dn.untr.des)`** statistically significant differentially expressed genes between **SEN_untreated and SR_untreated**. Among these genes, there are **`r nrow(up.untr.des)`** up-regulated and **`r nrow(dn.untr.des)`** in SEN untreated.

### SEN_treated vs. SR_treated

```{r DEtreat, cache = TRUE}

res = nbinomTest( cdsBlind, "sr_treated", "sen_treated")
```

```{r plotDEtreat}
plotMA(res)
hist(res$padj, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)

res <- getDEset(res, 0.05)

up.tr.des <- res[[1]]
dn.tr.des <- res[[2]]

head(up.tr.des)
head(dn.tr.des)
```

In total, there are **`r nrow(up.tr.des)+nrow(dn.tr.des)`** statistically significant differentially expressed genes between **SEN_treated and SR_untreated**. Among these genes, there are **`r nrow(up.tr.des)`** up-regulated and **`r nrow(dn.tr.des)`** in SEN untreated.

## Calling Differential Expression with NOISeq

### SR_treated vs. SR_untreated

```{r NOISeqSR, message = FALSE, results = "hide", cache = TRUE}
library(NOISeq)
set.seed(12345)

runNOI <- function(condition){
    facs <- groups[grep(condition, groups, perl = TRUE)]
    facs <- data.frame(Treatment = factor(facs), sampleRun = factor(facs))
    noiData = tmm[, grep(condition, colnames(tmm), perl = TRUE)]
    rownames(noiData) = tmm.df$hgnc_symbol
    noiData <- readData(noiData, factors=facs)
    res <- noiseq(noiData, factor="Treatment", k=NULL, norm="n", pnr=0.2, v=0.02, nss=5, replicates="no")
}

res <- runNOI("sr")
```

```{r message = FALSE}
library(NOISeq)
q.cut = 0.95
DE.plot(res, q=q.cut, graphic="MD")
prob_val <- res@results[[1]]$prob
hist(prob_val, breaks = 100, col = "steelblue", main = "", xlab = "Probability of DE")
abline(v = quantile(prob_val, q.cut, na.rm = TRUE), col = "red", lty = 2)

up.sr.noi <- degenes(res, q=q.cut, M="up")
dn.sr.noi <- degenes(res, q=q.cut, M="down")
up.sr.noi.df <- data.frame(id = rownames(up.sr.noi), up.sr.noi)
dn.sr.noi.df <- data.frame(id = rownames(dn.sr.noi), dn.sr.noi)
```

### SEN_treated vs. SEN_untreated

```{r NOISeqSEN, results = "hide", cache = TRUE}

res <- runNOI("sen")
```

```{r}
library(NOISeq)

DE.plot(res, q=q.cut, graphic="MD")
prob_val <- res@results[[1]]$prob
hist(prob_val, breaks = 100, col = "steelblue", main = "", xlab = "Probability of DE")
abline(v = quantile(prob_val, q.cut, na.rm = TRUE), col = "red", lty = 2)

up.sen.noi <- degenes(res, q=q.cut, M="up")
dn.sen.noi <- degenes(res, q=q.cut, M="down")
up.sen.noi.df <- data.frame(id = rownames(up.sen.noi), up.sen.noi)
dn.sen.noi.df <- data.frame(id = rownames(dn.sen.noi), dn.sen.noi)
```

### SEN_untreated vs. SR_untreated

```{r NOISeqUN, results = "hide", cache = TRUE}

res <- runNOI("sen_untreated|sr_untreated")
```

```{r}
library(NOISeq)

DE.plot(res, q=q.cut, graphic="MD")
prob_val <- res@results[[1]]$prob
hist(prob_val, breaks = 100, col = "steelblue", main = "", xlab = "Probability of DE")
abline(v = quantile(prob_val, q.cut, na.rm = TRUE), col = "red", lty = 2)

up.untr.noi <- degenes(res, q=q.cut, M="down") #Up in SEN
dn.untr.noi <- degenes(res, q=q.cut, M="up") #Down in SEN
up.untr.noi.df <- data.frame(id = rownames(up.untr.noi), up.untr.noi)
dn.untr.noi.df <- data.frame(id = rownames(dn.untr.noi), dn.untr.noi)
```

### SEN_treated vs. SR_treated

```{r NOISeqTR, results = "hide", cache = TRUE}

res <- runNOI("sen_treated|sr_treated")
```

```{r}
library(NOISeq)

DE.plot(res, q=q.cut, graphic="MD")
prob_val <- res@results[[1]]$prob
hist(prob_val, breaks = 100, col = "steelblue", main = "", xlab = "Probability of DE")
abline(v = quantile(prob_val, q.cut, na.rm = TRUE), col = "red", lty = 2)

up.tr.noi <- degenes(res, q=q.cut, M="down") #Up in SEN
dn.tr.noi <- degenes(res, q=q.cut, M="up") #Down in SEN
up.tr.noi.df <- data.frame(id = rownames(up.tr.noi), up.tr.noi)
dn.tr.noi.df <- data.frame(id = rownames(dn.tr.noi), dn.tr.noi)
```


## Calling Differential Expression with DEGSeq

### SR_treated vs. SR_untreated

```{r DEGSeqSR, results = "hide", cache = TRUE}
library(DEGseq)


runDEG <- function(cond1, cond2){
    set.seed(12345)
    par(mfrow = c(3, 2), mar = rep(2, 4))
    gem1 = as.matrix(tmm.df[, c(grep("hgnc", colnames(tmm.df)), grep(cond1, colnames(tmm.df)))])

    gem2 = as.matrix(tmm.df[, c(grep("hgnc", colnames(tmm.df)), grep(cond2, colnames(tmm.df)))])
    
    outDir = paste(cond1,"_", cond2, sep = "")
    DEGexp(geneExpMatrix1 = gem1, 
           geneCol1 = 1, 
           expCol1 = 2, 
           groupLabel1 = cond1, 
           geneExpMatrix2 = gem2, 
           geneCol2 = 1, 
           expCol2 = 2, 
           groupLabel2 = cond2,
           method = "MARS",
           outputDir = outDir)
    res <- read.delim(paste("./", outDir, "/output_score.txt", sep = ""),
                      header = TRUE)
    colnames(res) <- gsub("value1", cond1, colnames(res))
    colnames(res) <- gsub("value2", cond2, colnames(res))
    colnames(res) <- gsub("q.value.Benjamini.et.al..1995.", "q.value", colnames(res))
    colnames(res) <- gsub("log2.Fold_change.", "log2.Fold_change", colnames(res))
    return(res)
}


res <- runDEG("sr_treated", "sr_untreated")
```

```{r}
getPath <- function(cond1, cond2){
    outDir = paste(cond1,"_", cond2, sep = "")
    pic <- paste("./", outDir, "/output/result.png", sep = "")
}

pic <- getPath("sr_treated", "sr_untreated")
```

<img src="./`r pic`" />

```{r DEGSeqSRres}
res1 <- res[complete.cases(res), ]
up.sr.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change > 0, ]
dn.sr.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change < 0, ]
hist(res1$q.value, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)
```

### SEN_treated vs. SEN_untreated

```{r DEGSeqSEN, results = "hide", cache = TRUE}


res <- runDEG("sen_treated", "sen_untreated")
```

```{r}
pic <- getPath("sen_treated", "sen_untreated")
```

<img src="./`r pic`" />

```{r DEGSeqSENres}
res1 <- res[complete.cases(res), ]
up.sen.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change > 0, ]
dn.sen.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change < 0, ]
hist(res1$q.value, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)
```

### SEN_untreated vs. SR_untreated

```{r DEGSeqUntr, cache = TRUE, results="hide"}


res <- runDEG("sen_untreated", "sr_untreated")
```

```{r}
pic <- getPath("sen_untreated", "sr_untreated")
```

<img src="./`r pic`" />

```{r DEGSeqUntrRes}
res1 <- res[complete.cases(res), ]
up.untr.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change > 0, ]
dn.untr.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change < 0, ]
hist(res1$q.value, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)
```

### SEN_treated vs. SR_treated

```{r DEGSeqTreat, cache = TRUE, results="hide"}


res <- runDEG("sen_treated", "sr_treated")
```

```{r}
pic <- getPath("sen_treated", "sr_treated")
```

<img src="./`r pic`" />

```{r DEGSeqTreatRes}
res1 <- res[complete.cases(res), ]
up.tr.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change > 0, ]
dn.tr.deg <- res1[res1$q.value < 0.05 & res1$log2.Fold_change < 0, ]
hist(res1$q.value, breaks = 100, col = "steelblue", main = "", xlab = "Adjusted p-value")
abline(v=0.05, col = "red", lty = 2)
```

## Compare Differential Expression Methods

###Compare DESeq and NOISeq 

```{r results = "hide"}
library(VennDiagram)
library(RColorBrewer)
vennGen <- function(name1, up.m1, dn.m1, id1, name2, up.m2, dn.m2, id2){
    m1.up <- up.m1[, id1]
    m1.dn <- dn.m1[, id1]
    m2.up <- up.m2[, id2]
    m2.dn <- dn.m2[, id2]
    ##Remove NA
    m1.up <- m1.up[!is.na(m1.up)]
    m1.dn <- m1.dn[!is.na(m1.dn)]
    m2.up <- m2.up[!is.na(m2.up)]
    m2.dn <- m2.dn[!is.na(m2.dn)]
    
    res <- list( m1.up, m1.dn, m2.up, m2.dn)
    names(res) = c(paste(name1, "_up", sep = ""),
                   paste(name1, "_down", sep = ""),
                   paste(name2, "_up", sep = ""),
                   paste(name2, "_down", sep = ""))
    return(res)
}

x <- vennGen("DESeq", up.sr.des, dn.sr.des,"hgnc_symbol", "NOISeq", up.sr.noi.df, dn.sr.noi.df, "id")
venn.diagram(x, "sr_DESeq_NOISeq.png", imagetype = "png", fill = brewer.pal(4, "BuGn"))

x <- vennGen("DESeq", up.sen.des, dn.sen.des, "hgnc_symbol", "NOISeq", up.sen.noi.df, dn.sen.noi.df, "id")
venn.diagram(x, "sen_DESeq_NOISeq.png", imagetype = "png", fill = brewer.pal(4, "OrRd"))

x <- vennGen("DESeq", up.untr.des, dn.untr.des, "hgnc_symbol", "NOISeq", up.untr.noi.df, dn.untr.noi.df, "id")
venn.diagram(x, "untr_DESeq_NOISeq.png", imagetype = "png", fill = brewer.pal(4, "BuPu"))

x <- vennGen("DESeq", up.tr.des, dn.tr.des, "hgnc_symbol", "NOISeq", up.tr.noi.df, dn.tr.noi.df, "id")
venn.diagram(x, "tr_DESeq_NOISeq.png", imagetype = "png", fill = brewer.pal(4, "Blues"))

```

####SR DESeq vs. SR NOISeq

<img src="sr_DESeq_NOISeq.png" width=500 height=500 />

####SEN DESeq vs. SEN NOISeq

<img src="sen_DESeq_NOISeq.png" width=500 height=500 />

####Untreated DESeq vs. Untreated NOISeq

<img src="untr_DESeq_NOISeq.png" width=500 height=500 />

####Treated DESeq vs. Treated NOISeq

<img src="tr_DESeq_NOISeq.png" width=500 height=500 />


###Compare NOISeq and DEGseq
```{r results = "hide"}
x <- vennGen("DEGseq", up.sr.deg, dn.sr.deg, "GeneNames", "NOISeq", up.sr.noi.df, dn.sr.noi.df, "id")
venn.diagram(x, "sr_NOISeq_DEGseq.png", imagetype = "png", fill = brewer.pal(4, "BuGn"))

x <- vennGen("DEGseq", up.sen.deg, dn.sen.deg, "GeneNames", "NOISeq", up.sen.noi.df, dn.sen.noi.df, "id")
venn.diagram(x, "sen_NOISeq_DEGseq.png", imagetype = "png", fill = brewer.pal(4, "OrRd"))

x <- vennGen("DEGseq", up.untr.deg, dn.untr.deg, "GeneNames", "NOISeq", up.untr.noi.df, dn.untr.noi.df, "id")
venn.diagram(x, "untr_NOISeq_DEGseq.png", imagetype = "png", fill = brewer.pal(4, "BuPu"))

x <- vennGen("DEGseq", up.tr.deg, dn.tr.deg, "GeneNames", "NOISeq", up.tr.noi.df, dn.tr.noi.df, "id")
venn.diagram(x, "tr_NOISeq_DEGseq.png", imagetype = "png", fill = brewer.pal(4, "Blues"))
```

####SR NOISeq vs. SR DEGseq

<img src="sr_NOISeq_DEGseq.png" width=500 height=500 />

####SEN NOISeq vs. SEN DEGseq

<img src="sen_NOISeq_DEGseq.png" width=500 height=500 />

####Untreated NOISeq vs. Untreated DEGseq

<img src="untr_NOISeq_DEGseq.png" width=500 height=500 />

####Treated NOISeq vs. Treated DEGseq

<img src="tr_NOISeq_DEGseq.png" width=500 height=500 />

<!--
### Validate Known DE Genes for DESeq, NOISeq and DEGseq
-->

```{r validateKownDEG}
vdeg = c("GAPDH", "ACTB", "RPL13A", "C1orf43", "CHMP2A","EMC7", "SMARCA5", "HOXA1", "H2AFY", "IL2RA", "IL2RB", "IL2RG", "STAT5A", "STAT5B")
names(vdeg) = c("GAPDH", "ACTB", "RPL13A", "C1orf43", "CHMP2A","EMC7", "SMARCA5", "HOXA1", "H2AFY", "IL2RA", "IL2RB", "IL2RG", "STAT5A", "STAT5B")

vdegid = c("NM_000417", "NM_000878", "NM_000206")

delist <- list(up.sr.des = up.sr.des$hgnc_symbol, 
               dn.sr.des = dn.sr.des$hgnc_symbol,
               up.sr.noi = rownames(up.sr.noi), 
               dn.sr.noi = rownames(dn.sr.noi), 
               up.sr.deg = up.sr.deg$GeneNames, 
               dn.sr.deg = dn.sr.deg$GeneNames,
               up.sen.des = up.sen.des$hgnc_symbol, 
               dn.sen.des = dn.sen.des$hgnc_symbol,
               up.sen.noi = rownames(up.sen.noi), 
               dn.sen.noi = rownames(dn.sen.noi), 
               up.sen.deg = up.sen.deg$GeneNames, 
               dn.sen.deg = dn.sen.deg$GeneNames,
               up.tr.des = up.tr.des$hgnc_symbol, 
               dn.tr.des = dn.tr.des$hgnc_symbol,
               up.tr.noi = rownames(up.tr.noi), 
               dn.tr.noi = rownames(dn.tr.noi), 
               up.tr.deg = up.tr.deg$GeneNames, 
               dn.tr.deg = dn.tr.deg$GeneNames,
               up.untr.des = up.untr.des$hgnc_symbol, 
               dn.untr.des = dn.untr.des$hgnc_symbol,
               up.untr.noi = rownames(up.untr.noi), 
               dn.untr.noi = rownames(dn.untr.noi), 
               up.untr.deg = up.untr.deg$GeneNames, 
               dn.untr.deg = dn.untr.deg$GeneNames,
               counts.df = counts.df$hgnc_symbol)


res <- lapply(delist, function(x) unlist(lapply(vdeg, grep, x)))


```

# Expression Level of Individual Genes/Sequences

```{r IndividualGeneExpression, message = FALSE, fig.width = 10, fig.height = 4, warnings = FALSE}
library(bear)
#c6 <- c6.container
plotGenes <- function(x, data, dataset = "", groups){
    values <- data[data$hgnc_symbol == x, ]
    dataset <- ifelse(nchar(dataset) == 0, dataset, paste(" from ", dataset))
    if(nrow(values) == 0){
        print(paste("No data found for ", x))
    }
    else{
        cols <- grep("sr|sen|rej", colnames(values), perl = TRUE)
        
        values <- as.numeric(unlist(values)[cols])
        values.df <- data.frame(values, groups)
        sum.var <- summarySE(values.df, measurevar = "values", groupvars = "groups")
        
        g <- ggplot(sum.var, aes(x = factor(groups), y = values, fill = groups)) +
            geom_bar(stat = "identity") +
            geom_errorbar(aes(ymin = values - se, ymax = values + se),
                          size = 0.2,
                          width = 0.2) +
            scale_fill_manual(values = c6) +
            labs(title = paste(x, dataset), x = "", y = "Counts") + 
        theme_bw() +
        theme(panel.grid.major = element_blank(), 
              legend.position = "none",
              axis.text.x = element_text(angle = 45, hjust = 1))
        return(g)
    }
}



multiplot <- function(..., plotlist = NULL, file, cols = 1, layout = NULL) {
      require(grid)
    
      plots <- c(list(...), plotlist)
    
      numPlots = length(plots)
    
      if (is.null(layout)) {
        layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                        ncol = cols, nrow = ceiling(numPlots/cols))
      }
    
      if (numPlots == 1) {
        print(plots[[1]])
    
      } else {
        grid.newpage()
        pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
        for (i in 1:numPlots) {
          matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
    
          print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                          layout.pos.col = matchidx$col))
        }
      }
}

old.hk <- c("GAPDH", "RPL13A", "ACTB")
new.hk <- c("C1orf43", "CHMP2A", "EMC7")
others <- c("SMARCA5","HOXA1","H2AFY", "STAT5A", "STAT5B")

datasets <- list(counts.df, tmm.df)
setnames <- c("Raw", "TMM")
names(datasets) <- setnames

plotMore <- function(datasets, setnames, genes){
    res <- lapply(genes, plotGenes, datasets, setnames, groups = groups)
    multiplot(plotlist = res, cols = 3 )
}


#Map(plotMore, datasets, setnames, rep(list(old.hk), length(datasets)))
```

```{r message = FALSE, warning = FALSE, fig.height = 6, fig.wdith = 14}
Map(plotMore, datasets, setnames, rep(list(others), length(datasets)))
```


# Final Results

## Downloadable Files

<!--

## DE Genes from DESeq

```{r}
write.csv(up.sr.des, "up.sr_tr.DESeq.csv", quote = FALSE)
write.csv(dn.sr.des, "dn.sr_tr.DESeq.csv", quote = FALSE)

write.csv(up.sen.des, "up.sen_tr.DESeq.csv", quote = FALSE)
write.csv(dn.sen.des, "dn.sen_tr.DESeq.csv", quote = FALSE)

write.csv(up.untr.des, "up.untr_sen.DESeq.csv", quote = FALSE)
write.csv(dn.untr.des, "dn.untr_sen.DESeq.csv", quote = FALSE)

write.csv(up.tr.des, "up.tr_sen.DESeq.csv", quote = FALSE)
write.csv(dn.tr.des, "dn.tr_sen.DESeq.csv", quote = FALSE)
```

### Compare DESeq Results from "Drop Small Libs" and "Combine Libs"


up.sr.des.drop <- read.csv("up.sr_tr.DESeq.drop.csv", header = TRUE)
dn.sr.des.drop <- read.csv("dn.sr_tr.DESeq.drop.csv", header = TRUE)

up.sen.des.drop <- read.csv("up.sen_tr.DESeq.drop.csv", header = TRUE)
dn.sen.des.drop <- read.csv("dn.sen_tr.DESeq.drop.csv", header = TRUE)

up.untr.des.drop <- read.csv("up.untr_sen.DESeq.drop.csv", header = TRUE)
dn.untr.des.drop <- read.csv("dn.untr_sen.DESeq.drop.csv", header = TRUE)

up.tr.des.drop <- read.csv("up.tr_sen.DESeq.drop.csv", header = TRUE)
dn.tr.des.drop <- read.csv("dn.tr_sen.DESeq.drop.csv", header = TRUE)

x <- vennGen("Drop", up.sr.des.drop, dn.sr.des.drop, "hgnc_symbol", "Combine", up.sr.des, dn.sr.des, "hgnc_symbol")
venn.diagram(x, "sr_Drop_vs_Combine.png", imagetype = "png", fill = brewer.pal(4, "BuGn"))

x <- vennGen("Drop", up.sen.des.drop, dn.sen.des.drop, "hgnc_symbol", "Combine", up.sen.des, dn.sen.des, "hgnc_symbol")
venn.diagram(x, "sen_Drop_vs_Combine.png", imagetype = "png", fill = brewer.pal(4, "OrRd"))

x <- vennGen("Drop", up.untr.des.drop, dn.untr.des.drop, "hgnc_symbol", "Combine", up.untr.des, dn.untr.des, "hgnc_symbol")
venn.diagram(x, "untr_Drop_vs_Combine.png", imagetype = "png", fill = brewer.pal(4, "BuPu"))

x <- vennGen("Drop", up.tr.des.drop, dn.tr.des.drop, "hgnc_symbol", "Combine", up.tr.des, dn.tr.des, "hgnc_symbol")
venn.diagram(x, "tr_Drop_vs_Combine.png", imagetype = "png", fill = brewer.pal(4, "Blues"))


-->


[Differentially Expressed Protein-coding Genes from GFOLD][f001]

## Visualzation in UCSC Genome Browser

Tracks are prepared to be be loaded to UCSC Genome Browser for visualization. Click [here][linkout] to browse.

**Description of Tracks**

Short Label | Description | Track Type
---- | ---- | ----
Normalized | Normalized expression of SR treated, SR untreated, SEN treated, SEN untreated | Multiple Wiggle
Raw | Raw/Not normalized expression of SR treated, SR untreated, SEN treated, SEN untreated | Multiple Wiggle
SEN TreatedBam | Reads mapped to SEN treated | bam
SEN TrScSingle | Normalized expression of SEN treated | Wiggle
SEN TrSingle | Raw/Not normalized expression of SEN treated | Wiggle
SEN UntreatedBam | Reads mapped to SEN untreated | bam
SEN UntrScSingle | Normalized expression of SEN untreated | Wiggle
SEN UntrSingle | Raw/Not normalized expression of SEN untreated | Wiggle
SR TreatedBam | Reads mapped to SR treated | bam
SR TrScSingle | Normalized expression of SR treated | Wiggle
SR TrSingle | Raw/Not normalized expression of SR treated | Wiggle
SR UntreatedBam | Reads mapped to SR untreated | bam
SR UntrScSingle | Normalized expression of SR untreated | Wiggle
SR UntrSingle | Raw/Not normalized expression of SR untreated | Wiggle


####Color used

colors in the plots `r c6`

colors in the tracks:

```{r}
library(grDevices)
tc6 = brewer.pal(6, "Pastel1")

#tc6 <- tc6[c(5, 1, 4, 8, 7, 11)]

col2rgb(tc6)
```

## Meeting Notes

[20140523][002]

[20140606][003]

[20140828][004]

[20140918][005]


[img01]:Kazak_Overall_experimental_design.jpg

[f001]:./data/kazak.DE.pcg.GFOLD.xlsx

[001]:http://www.lifetechnologies.com/order/catalog/product/4456740
[002]:./slides/20140523.Overall_analytical_design.pdf
[003]:./slides/20140606.ERCC.pdf
[004]:./slides/20140828.Normalization_of_kazak_RNA-seq_data.pdf
[005]:./slides/20140918.Comapring_DE_Methods.pdf

[linkout]:http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&hubUrl=http://jordan.biology.gatech.edu/lulu/trackHub/hub.txt
