```{r echo=FALSE}
library(knitr)
opts_chunk$set(echo=FALSE)
```

---
title: "IL-2 Influences of the Transcriptional Activity in Human MSC"
author: "Lu Wang"
date: "Monday, July 21, 2014"
output: html_document
---
## Experimental Design

![Experimental Design][img01]

## Libraries and Raw Data

In this analysis, we started with `.bam` files containing RNA-seq reads aligned to hg19. The first metadata file is the `libs` file which contains the name and total mapped reads of all libraries.

```{r cache=TRUE}
libs <- read.delim("./kazak_data/tmap_lib_size.tsv", header=TRUE,sep="\t")
```

```{r colors}
library(RColorBrewer)
c6 <- brewer.pal(6, "Paired")
c6 <- c6[c(2, 1, 6, 5, 4, 3)]

pal1 <- colorRampPalette(c6[1:2])
pal2 <- colorRampPalette(c6[3:4])
pal3 <- colorRampPalette(c6[5:6])

c15 <- c(pal1(7)[c(1:2, 5:7)], pal2(6)[c(1:2, 5:6)], pal3(8)[c(1:3, 6:8)])
```

In this data set, we have **`r nrow(libs)`** sequencing runs that correspond to **six** different RNA libraries/conditions:

1. SR
2. SR with IL-2 treatment
3. SEN
4. SEN with IL-2 treatment
5. REJ
6. REJ with IL-2 treatment

Each file contains mapped reads were pooled together, compared against RefSeq gene model(`RefSeq.bed` file downloaded from UCSC genome browser) and generated a counts matrix using `bedtools`. 

## Distribution of Raw Counts

The raw counts for each transcript is calculated by couting the number of reads overlapped with the **exons** of both protein-coding and non-coding **RefSeq transcripts**.

```{r readCountsMatrix, cache=TRUE}
tb <- read.table("./kazak_data/counts_matrix.txt",header=FALSE)
counts <- as.matrix(tb[,2:16])
rownames(counts) <- tb[,1]
rm(tb)
colnames(counts) <- rownames(libs)
counts <- counts[order(rownames(counts)),]
head(counts, n=2)
```


```{r CountsDensity}
library(reshape)
library(ggplot2)

plotDistribution <- function(data, title = ""){
    par(mar=c(8.1,6.1,4.1,2.1))
    par(las=2)
    boxplot(log2(data.frame(data)), col = c15, names = (colnames(data)), main = paste(title))
    data.df <- data.frame(id = rownames(data), data, stringsAsFactors = FALSE)
    mdf <- melt(data.df, id = "id", variable_name = "condition")

    ggplot(mdf, aes(log2(value))) +
        geom_density(aes(color = condition)) + 
        title(main = paste(title), xlab = "") + 
        scale_colour_manual(values = c15) +
        theme_bw() +
        theme(panel.grid.major = element_blank())
}

plotDistribution(counts+1, "Distribution of Read Counts")
```

In this analysis, **technical replicates for each condition were combined** to increase the total signal. We started with **`r nrow(counts)`** RefSeq transcript annotations.

```{r filterLibs, fig.width = 6}
c15 <- c6
C15 <- c15[1:4]
sr_tr = grep("sr_treated", colnames(counts))
sr_un = grep("sr_untreated", colnames(counts))
sen_tr = grep("sen_treated", colnames(counts))
sen_un = grep("sen_untreated", colnames(counts))
rej_tr = grep("rej_treated", colnames(counts))
rej_un = grep("rej_untreated", colnames(counts))


counts <- data.frame( sr_untreated = apply(counts[, sr_un], 1, sum)
                      ,sr_treated = apply(counts[, sr_tr], 1, sum)
                      ,sen_untreated = apply(counts[, sen_un], 1, sum)
                      ,sen_treated = apply(counts[, sen_tr], 1, sum)
                      #,rej_untreated = apply(counts[, rej_un], 1, sum)
                      #,rej_treated = apply(counts[, rej_tr], 1, sum)
                    )
head(counts, 2)
plotDistribution(counts+1, "Distribution of Read Counts")
```

```{r libSizes}
barplot(colSums(counts), col = c6)
```



## ERCC Counts

The ERCC counts for each library were downloaded from the **Ion Proton** server with the help from James. The distribution of ERCC counts is shown below:

```{r readERCC, fig.width = 6}
ercc <- read.table("./kazak_data/ercc_counts.txt", header = TRUE)

ercc <- data.frame( sr_untreated = apply(ercc[, sr_un], 1, sum)
                    , sr_treated = apply(ercc[, sr_tr], 1, sum)
                    , sen_untreated = apply(ercc[, sen_un], 1, sum)
                    , sen_treated = apply(ercc[, sen_tr], 1, sum)
                    #, rej_untreated = apply(ercc[, rej_un], 1, sum)
                    #, rej_treated = apply(ercc[, rej_tr], 1, sum)
                  )
head(ercc, 2)

plotDistribution(ercc+1, "Distribution of ERCC Counts")
```

In general, the distribution of **`r nrow(ercc)`** ERCC counts follows the trend of total read counts.

## Determining Lower Limit of Detection

The true counts of the ERCC sequences were obtaind from the `ERCC_Controls_Analysis.txt` file downloaded from [manufacturer's website][001]

```{r downloadERCC, cache = TRUE}
con = "http://tools.lifetechnologies.com/downloads/ERCC_Controls_Analysis.txt"
true.ercc <- read.delim(con, header = TRUE)
```

The scatterplots below compared the observed ERCC counts versus the theoretical number of molecules from the manufacturer's documentation.

```{r reformatERCC}
true.ercc <- true.ercc[, c(2, 4)]
true.ercc <- true.ercc[order(true.ercc$ERCC.ID), ]

head(true.ercc)
colnames(true.ercc) = c("ercc.id", "mix1.attmoles")
```


```{r obsERCCvsTrue, fig.height = 6, fig.width = 6}
library(reshape)
library(lattice)

low.cut <- 3
lib.cut <- 2

ercc.df <- cbind(ercc.id = rownames(ercc), ercc, stringsAsFactors = FALSE)

erccOvsE <- function(data, title = ""){
    temp <- merge(data, true.ercc, by.x = colnames(data)[grep("id", colnames(data))], by.y = "ercc.id")
    
    mdf <- melt(temp[, c(1, grep("sr|sen|rej", colnames(temp)))], id = colnames(data)[grep("id", colnames(data))], variable_name = "sample")
    
    mdf <- cbind(mdf, expected = rep(temp$mix1.attmoles, ncol(ercc)))
    
    mdf$sample <- ordered(mdf$sample)
    
    xyplot( log2(value + 1) ~ log2(expected+1) | sample, data = mdf, 
            layout = c(2,2), 
            as.table = TRUE, 
            pch = 20, 
            col = rgb(70, 130, 180, 100, maxColorValue = 255), 
            par.strip.text = list(cex = 0.7), 
            xlim = c(0, 15),
            ylim = c(0, 15),
            xlab = expression(log[2]*"Expected Counts"), 
            ylab = expression(log[2]*"Observed Counts"), 
            main = paste(title), 
            panel = function(x, y, ...) {
                panel.xyplot(x, y, ...)
                panel.abline(h = log2(low.cut+1), lty = 2)
                panel.lmline(x, y, col.line = rgb(0,0,0,0.2))
    })
}
erccOvsE(ercc.df, "ERCC Dose Response")
```

The **dashed line** in panels above shows the empirical cutoff for low read count sequences. The ERCC sequences with more than **`r low.cut`** counts in each library demonstrated linear relationship between the expected and observed read counts.

```{r plotPairwiseCorr, fig.height = 9, fig.width = 12}
plotPairs <- function(data, title = ""){
    pairs(log2(data), col = rgb(0,0.5,1,0.2), pch = 16, main = paste(title))
}
```


# Normalization

```{r exonLengths, cache=TRUE}
exon.lengths <- t(read.delim("./kazak_data/exon_lengths.tsv",header=TRUE,sep="\t"))
```

The exon length of each RefSeq transcripts were calculated using the `RefSeq.gtf` file downloaded from UCSC genome browser. 

```{r getGeneLengths}
counts.df <- data.frame(id = rownames(counts), counts, stringsAsFactors = FALSE)
lengths.df <- data.frame(id = rownames(exon.lengths), exon.lengths)
mdf <- merge(counts.df, lengths.df, by.x = "id", by.y = "id")
counts.df <- mdf
head(counts.df, 2)
```

For ERCC Spike-Ins, the lengths of the ERCC sequences were obtaind from the `ERCC92.gtf` file downloaded from [manufacturer's website][001]

```{r getERCClength}
tb <- read.table("http://tools.invitrogen.com/downloads/ERCC92.gtf", header = FALSE)
tb <- tb[, c(1, 5)]
colnames(tb) <- c("id", "length")
mdf <- merge(ercc.df, tb, by.x = "ercc.id", by.y = "id" )
ercc.df <- mdf
head(ercc.df, 2)
```


```{r RefSeq2HGNC_symbol, cache = TRUE}
library(biomaRt)


pcg <- grep("NM_", counts.df$id)
nc <- grep("NR_", counts.df$id)

pcg.id <- counts.df$id[pcg]
nc.id <- counts.df$id[nc]

getSymbol <- function(ids, id.type, attributes){
    ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
    res <- getBM( attributes = c(paste(id.type), paste(attributes)), filters = id.type, values = ids, mart = ensembl )
}


res1 <- getSymbol(pcg.id, "refseq_mrna", attributes = "hgnc_symbol")
res2 <- getSymbol(nc.id, "refseq_ncrna", attributes = "hgnc_symbol")
```

```{r cleanHGNC_symbol}
cleanBiomartRes <- function(data, id.type, attribute){
    # Remove nchar=0 empty results
    data[, attribute] <- ifelse(nchar(data[, paste(attribute)]) == 0, NA, data[, attribute])
    # Sort and remove duplicated id
    data <- data[order(data[, id.type], data[, attribute]), ]
    data <- data[!duplicated(data[, id.type]), ]
}
res1 <- cleanBiomartRes(res1, "refseq_mrna", "hgnc_symbol")
res2 <- cleanBiomartRes(res2, "refseq_ncrna", "hgnc_symbol")
```

```{r assembleHGNC_symbol}
#Assemble results for mRNA
pcg <- merge(counts.df[pcg, ], res1, by.x = "id", by.y = "refseq_mrna", all.x = TRUE)

fillMissingSymbol <- function(data){
    data$hgnc_symbol <- ifelse(is.na(data$hgnc_symbol), data$id, data$hgnc_symbol)
    return(data)
}
pcg <- fillMissingSymbol(pcg)
#rm(res1)

#Assemble results for ncRNA
nc <- merge(counts.df[nc, ], res2, by.x = "id", by.y = "refseq_ncrna", all.x = TRUE)
nc <- fillMissingSymbol(nc)
#rm(res2)
#Assemble mRNA and ncRNA
counts.df <- rbind(pcg, nc)
rm(pcg)
rm(nc)

ercc.df <- data.frame(ercc.df, hgnc_symbol = ercc.df$ercc.id, stringsAsFactors = FALSE)
colnames(ercc.df) <- gsub("ercc.id", "id", colnames(ercc.df))
colnames(counts.df) <- gsub("Length", "length", colnames(counts.df))
counts.df <- rbind(ercc.df, counts.df)

dataCols <- grep("sr|sen|rej", colnames(counts.df), perl = TRUE)
annoCols <- grep("id|symbol|length", colnames(counts.df), perl = TRUE)

groups <- colnames(counts.df)[dataCols]
groups <- gsub("_[0-9]+", "", groups)
groups <- factor(groups, levels = unique(groups))
```

```{r message = FALSE}
library(edgeR)
dge <- DGEList(counts = counts.df[, dataCols], group = groups, genes = counts.df[, annoCols])
```

## Filtering Low Expression Sequences


```{r FilterCounts, fig.width = 6}
#Filter Low Count Seqeunces
#keep <- apply(dge$counts, 1, min) > 3
keep <- rowSums(cpm(dge) > 1) >= 1
dge <- dge[keep, ]

plotDistribution(dge$counts, "Distribution of Filtered Counts")

counts.df <- data.frame(dge$counts, dge$genes)
```

After filtering, **`r sum(keep)`** transcripts were kept for downstream analysis.

## Expression of Genes
Since different RefSeq transcripts for the same gene symbol shares most of their reads, therefore, among the **`r nrow(counts.df)`** transcripts, the transcript with highest overall count for each gene symbol were used for that particular gene symbol. 

```{r getGeneExpression}
o <- order(rowSums(dge$counts), decreasing = TRUE)
dge <- dge[o, ]
dge <- dge[! duplicated(dge$genes$hgnc_symbol), ]

dge$samples$lib.size <- colSums(dge$counts)
```


## Normalize against beta-actin(ACTB)

```{r}
actb <- counts.df[counts.df$hgnc_symbol == "ACTB", ]
actb.val <- unlist(actb[, grep("sr|sen|rej", colnames(actb), perl = TRUE)])
actb.sc <- actb.val/mean(actb.val)
```

```{r}
val <- grep("sr|sen|rej", colnames(counts.df), perl = TRUE)
nonval <- grep("sr|sen|rej", colnames(counts.df), perl = TRUE, invert = TRUE)
actb <- t(t(counts.df[, val])/actb.sc)
actb.df <- data.frame(actb, counts.df[, nonval])
```

```{r TMMDensity, fig.width = 6}
library(reshape)
library(ggplot2)

plotDistribution(actb, "ACTB Normalized Gene Expression")

plotDistribution(actb[grep("ERCC-", actb.df$hgnc_symbol), ], "ACTB Normalized ERCC Sequences")
```


# Calling Differential Expression with GFOLD

```{r message = FALSE}
#remote.gfold = "http://jordan.biology.gatech.edu/lulu/data/kazak.DEgenes.normbyACTB.xlsx"
#dest.gfold = "./kazak_data/kazak.DEgenes.normbyACTB.xlsx"
#download.file(remote.gfold, dest.gfold, mode = "wb")
##Manuall converted to csv due to large .xlsx file size
sr.gfold = "./kazak_data/kazak.DEgenes.normbyACTB-sr.csv"
sen.gfold = "./kazak_data/kazak.DEgenes.normbyACTB-sen.csv"
sr.ws <- read.csv(sr.gfold, header = TRUE)
sen.ws <- read.csv(sen.gfold, header = TRUE)
```


## Input for GFOLD: ACTB + Gene Length Normalized Expression


```{r}
mdf <- merge(sr.ws, sen.ws, by.x = "RefSeq.ID", by.y = "RefSeq.ID")
colnames(mdf) <- gsub("\\.x", "\\.sr", colnames(mdf))
colnames(mdf) <- gsub("\\.y", "\\.sen", colnames(mdf))

gfold.df <- mdf[, grep("Gene.Symbol.sen", colnames(mdf), invert = TRUE)]
gfold.exprs <- grep("SR|SEN", colnames(gfold.df), perl = TRUE)
```

```{r gfoldDistribution}
#pdf("actb+length_distribution.pdf", height = 10, width = 10)
plotDistribution(gfold.df[, gfold.exprs], "Distribution of ACTB + Length Normalized Counts")
#dev.off()
```

## Comparing DE Genes in SR and SEN upon Treatment
```{r}
gfold.cut = 0.01
#gfold.cut = quantile(gfold.df$GFOLD.sr, 0.95) 
perc.de <- mean(gfold.df$GFOLD.sr > gfold.cut | gfold.df$GFOLD.sr < -gfold.cut)
```

```{r gfoldDE}
gfold.up.sr = gfold.df$GFOLD.sr > gfold.cut
gfold.up.sen = gfold.df$GFOLD.sen > gfold.cut
gfold.dn.sr = gfold.df$GFOLD.sr < -gfold.cut
gfold.dn.sen = gfold.df$GFOLD.sen < -gfold.cut


gfold.up.union = (gfold.df$GFOLD.sr > gfold.cut | gfold.df$GFOLD.sen > gfold.cut)
gfold.up.sr.u = (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen <= gfold.cut)
gfold.up.sen.u = (gfold.df$GFOLD.sr <= gfold.cut & gfold.df$GFOLD.sen > gfold.cut)
gfold.up.intersect = (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen > gfold.cut)


gfold.dn.union = (gfold.df$GFOLD.sr < -gfold.cut | gfold.df$GFOLD.sen < -gfold.cut)
gfold.dn.sr.u = (gfold.df$GFOLD.sr < -gfold.cut & gfold.df$GFOLD.sen >= -gfold.cut)
gfold.dn.sen.u = (gfold.df$GFOLD.sr >= -gfold.cut & gfold.df$GFOLD.sen < -gfold.cut)
gfold.dn.intersect = (gfold.df$GFOLD.sr < -gfold.cut & gfold.df$GFOLD.sen < -gfold.cut)
```

With the GFOLD score cutoff set at **`r gfold.cut`**, we have **`r perc.de*100`%** genes differentially expressed. Below are the results for differentially expressed genes:

Set Name | Number of Gene
---- | ----
Up-regulated genes in SR treated | `r sum(gfold.up.sr)`
Down-regulated genes in SR treated | `r sum(gfold.dn.sr)`
Up-regulated in SEN treated | `r sum(gfold.up.sen)`
Down-regulated in SEN treated | `r sum(gfold.dn.sen)`

```{r vennDiagram}
library(VennDiagram)

vennGen <- function(name1, up.m1, name2, up.m2, id){
    m1.up <- up.m1[, id]
    m2.up <- up.m2[, id]
    res <- list(m1.up, m2.up)
    names(res) = c(name1, name2)
    return(res)
}

x <- vennGen("SEN_up", gfold.df[gfold.up.sen, ], "SR_up", gfold.df[gfold.up.sr, ], "Gene.Symbol.sr")
venn.diagram(x, "sr_sen_up.png", imagetype = "png", fill = c6[5:6])

x <- vennGen("SR_down", gfold.df[gfold.dn.sr, ], "SEN_down", gfold.df[gfold.dn.sen, ], "Gene.Symbol.sr")
venn.diagram(x, "sr_sen_down.png", imagetype = "png", fill = c6[3:4])


vennGen <- function(name1, up.m1, name2, up.m2, name3, dn.m3, name4, dn.m4, id){
    m1.up <- up.m1[, id]
    m2.up <- up.m2[, id]
    m3.dn <- dn.m3[, id]
    m4.dn <- dn.m4[, id]
    res <- list(m1.up, m2.up, m3.dn, m4.dn)
    names(res) = c(name1, name2, name3, name4)
    return(res)
}

x <- vennGen("SR_up", gfold.df[gfold.up.sr, ], "SEN_up", gfold.df[gfold.up.sen, ], "SR_down", gfold.df[gfold.dn.sr, ], "SEN_down", gfold.df[gfold.dn.sen, ], "Gene.Symbol.sr")

venn.diagram(x, "sr_sen_up_down.png", imagetype = "png", fill = c6[c(5,6,3,4)])

```

<img src="sr_sen_up.png" width=400 height=400 />
<img src="sr_sen_down.png" width=400 height=400 />


## Hierachical Clustering

Hierarchical clustering was performed and heatmap was generated to visualize gene clusters. 8 heatmaps were generated for

Set Name | Number of Gene
---- | ----
Union of up-regulated genes in SR treated and SEN treated | `r sum(gfold.up.union)`
Genes up-regulated in SR treated only | `r sum(gfold.up.sr.u)`
Genes up-regulated in SEN treated only | `r sum(gfold.up.sen.u)`
Intersection of up-regulated genes in SR treated and SEN treated | `r sum(gfold.up.intersect)`
Union of down-regulated genes in SR treated and SEN treated | `r sum(gfold.dn.union)`
Genes down-regulated in SR treated only | `r sum(gfold.dn.sr.u)`
Genes down-regulated in SEN treated only | `r sum(gfold.dn.sen.u)`
Intersection of down-regulated genes in SR treated and SEN treated | `r sum(gfold.dn.intersect)`

```{r eval = FALSE}
##SEN GFOLD vs FC
g <- ggplot(gfold.df, 
            aes(log2(gfold.df$SR.treated/gfold.df$SR.untreated),
                gfold.df$GFOLD.sr))

g + geom_point(aes(col = gfold.df$GFOLD.sr), 
               size = 2 ) + 
    scale_color_gradient2(low = "green", mid = "grey", high = "red", midpoint = 0) + 
    coord_cartesian(xlim = c(-2, 2), ylim = c(-2, 2)) + 
    theme_bw() +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank()) +
    labs(title = "Up- and Down-regulated Genes in SR")

##SEN GFOLD vs FC
g <- ggplot(gfold.df, 
            aes(log2(gfold.df$SEN.treated/gfold.df$SEN.untreated),
                gfold.df$GFOLD.sen))

g + geom_point(aes(col = gfold.df$GFOLD.sen), 
               size = 2 ) + 
    scale_color_gradient2(low = "green", mid = "grey", high = "red", midpoint = 0) + 
    coord_cartesian(xlim = c(-4, 4), ylim = c(-4, 4)) + 
    theme_bw() +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank()) +
    labs(title = "Up- and Down-regulated Genes in SEN")
```


```{r hclust, eval = FALSE}
library(gplots)
hmcol = colorRampPalette(c("green", "darkgreen", "black","darkred", "red"))(100)
# plotHeatmap <- function(data, colIndex, rowIndex, title){
#     mat <- as.matrix(data[rowIndex, colIndex])
# 
#     zmat <- (mat - rowMeans(mat))/apply(mat, 1, sd)
#     
#     zmat1 <- zmat[, 1:2]
#     zmat2 <- zmat[, 3:4]
#     med1 <- median(zmat1)
#     med2 <- median(zmat2)
#     zmat1 = zmat1 - med1
#     zmat2 = zmat2 - med2
#     zmat = cbind(zmat1, zmat2)
#     
#     heatmap.2(zmat, 
#               col = hmcol,
#               Colv = FALSE,
#               trace = "none", 
#               scale = "none",
#               labRow = data[, grep("Symbol", colnames(data))],
#               margin = c(15, 6),
#               keysize = 1.5,
#               density.info = "none",
#               key.title = "z-score",
#               key.xlab = "z-score",
#               main = title)
# }


## Up Treated

# pdf("heatmaps.pdf", height = 10 , width = 5)
# plotHeatmap(gfold.df, gfold.exprs, gfold.up.sr.u, "SR Treated Up Only")
# plotHeatmap(gfold.df, gfold.exprs, gfold.up.sen.u, "SEN Treated Up Only")
# plotHeatmap(gfold.df, gfold.exprs, gfold.up.intersect, "Treated Up intersection")

## Down Treated

# plotHeatmap(gfold.df, gfold.exprs, gfold.dn.sr.u, "SR Treated Down Only")
# plotHeatmap(gfold.df, gfold.exprs, gfold.dn.sen.u, "SEN Treated Down Only")
# plotHeatmap(gfold.df, gfold.exprs, gfold.dn.intersect, "Treated Down Intersect")
# dev.off()


## TOP 500 or less
top.n = 500

plotHeatmap <- function(data, rowIndex, sort.by, title){
    data <- data[rowIndex, ]
    if( length(sort.by) > 1 ){
        data$key <- data[, sort.by[1]] + data[, sort.by[2]]
        if( length(grep("Up", title)) > 0 ){
            data <- data[order(data$key, decreasing = TRUE), ]
        }else{
            data <- data[order(data$key), ]
        }
    }else{
        if( length(grep("Up", title)) > 0 ){
            data <- data[order(data[, sort.by], decreasing = TRUE), ]
        }else{
            data <- data[order(data[, sort.by]), ]
        }
    }

    data <- data[1:min(nrow(data), 500), ]

    mat <- as.matrix(data[, gfold.exprs])

    zmat <- (mat - rowMeans(mat))/apply(mat, 1, sd)
    
    zmat1 <- zmat[, 1:2]
    zmat2 <- zmat[, 3:4]
    med1 <- median(zmat1)
    med2 <- median(zmat2)
    zmat1 = zmat1 - med1
    zmat2 = zmat2 - med2
    zmat = cbind(zmat1, zmat2)
    
    heatmap.2(zmat, 
              col = hmcol,
              Colv = FALSE,
              trace = "none", 
              scale = "none",
              labRow = data[, grep("Symbol", colnames(data))],
              margin = c(15, 6),
              keysize = 1.5,
              density.info = "none",
              key.title = "z-score",
              key.xlab = "z-score",
              main = title)
}

## Up Treated

#pdf("heatmaps.pdf", height = 10 , width = 5)
plotHeatmap(gfold.df, gfold.up.sr.u, grep("GFOLD.sr", colnames(gfold.df)), "SR Treated Up Only")
plotHeatmap(gfold.df, gfold.up.sen.u, grep("GFOLD.sen", colnames(gfold.df)), "SEN Treated Up Only")
plotHeatmap(gfold.df, gfold.up.intersect, grep("GFOLD", colnames(gfold.df)), "Treated Up intersection")

## Down Treated

plotHeatmap(gfold.df, gfold.dn.sr.u, grep("GFOLD.sr", colnames(gfold.df)), "SR Treated Down Only")
plotHeatmap(gfold.df, gfold.dn.sen.u, grep("GFOLD.sen", colnames(gfold.df)), "SEN Treated Down Only")
plotHeatmap(gfold.df, gfold.dn.intersect, grep("GFOLD", colnames(gfold.df)), "Treated Down Intersect")
#dev.off()
```

```{r plotSelectedHeatmap, eval=FALSE}
library(gplots)
hmcol = colorRampPalette(c("green", "black", "red"))(100)

## Selected gene from Victoria
geneset <- read.csv("./kazak_data/DATA_for_hitmaps_VL-Lulu.csv", header = T)

geneset.p <- split(geneset, as.factor(geneset$panel))
geneset.p.dfs <- lapply(geneset.p, function(x) res = merge(x, gfold.df, by.x = "symbol", by.y = "Gene.Symbol.sr"))


plotSelectedHeatmap <- function(data, title){
    mat <- as.matrix(data[, grep("treated|untreated", names(data), perl = T)])
    
    zmat <- (mat - rowMeans(mat))/apply(mat, 1, sd)
    #zmat = mat
    
    zmat1 <- zmat[, 1:2]
    zmat2 <- zmat[, 3:4]
    #med1 <- median(zmat1)
    #med2 <- median(zmat2)
    med1 <- rowMeans(zmat1)
    med2 <- rowMeans(zmat2)
    zmat1 = zmat1 - med1
    zmat2 = zmat2 - med2
    zmat = cbind(zmat1, zmat2)
    
    heatmap.2(zmat, 
              col = hmcol,
              Colv = FALSE,
              trace = "none", 
              scale = "none",
              labRow = data[, grep("gene", colnames(data))],
              margin = c(15, 6),
              keysize = 1.5,
              density.info = "none",
              key.title = "z-score",
              key.xlab = "z-score",
              main = title)
}

#pdf("selected-gene-heatmap1.pdf", height = 10, width = 6)
mapply(plotSelectedHeatmap, geneset.p.dfs, title = c("A", "B", "C", "D"))
#dev.off()


```



```{r sectionDensity, eval = FALSE}
library(fields)
up.sr.up.sen = (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen > gfold.cut)
up.sr.no.sen = (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen <= gfold.cut)
up.sr.dn.sen = (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen < -gfold.cut)
no.sr.dn.sen = (gfold.df$GFOLD.sr <= gfold.cut & gfold.df$GFOLD.sen < -gfold.cut)

#gfold.df$Cut.sr = with(gfold.df, 
#                       cut(GFOLD.sr, 
#                           c(min(GFOLD.sr), 
#                             -gfold.cut, 
#                             gfold.cut, 
#                             max(GFOLD.sr))))

gfold.df$color = as.factor(with(gfold.df, paste("SR", Regulation.sr, "SEN", Regulation.sen)))

table(gfold.df$color)

#pdf("gfold.density.pdf", height = 10, width = 12)

fudgeit <- function(){
  xm <- get('xm', envir = parent.frame(1))
  ym <- get('ym', envir = parent.frame(1))
  z  <- get('dens', envir = parent.frame(1))
  colramp <- get('colramp', parent.frame(1))
  image.plot(xm,ym,z, col = colramp(256), legend.only = T, add =F)
}

par(mar = c(5,4,4,5) + .1)

smoothScatter(gfold.df$GFOLD.sr, gfold.df$GFOLD.sen, xlim = c(-2, 2), ylim = c(-2, 2), xlab = "GFOLD score SR", ylab = "GFOLD score SEN", postPlotHook = fudgeit)
#abline(h =  gfold.cut, lty = 3, lwd = 1, col = rgb(0, 0, 0, 0.5))
#abline(h = -gfold.cut, lty = 3, lwd = 1, col = rgb(0, 0, 0, 0.5))
#abline(v =  gfold.cut, lty = 3, lwd = 1, col = rgb(0, 0, 0, 0.5))
#abline(v = -gfold.cut, lty = 3, lwd = 1, col = rgb(0, 0, 0, 0.5))


sel <- (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen > gfold.cut) | (gfold.df$GFOLD.sr < -gfold.cut & gfold.df$GFOLD.sen < -gfold.cut)

g <- ggplot(gfold.df[sel, ], aes(GFOLD.sr, GFOLD.sen)) 
g + geom_point(aes(col = GFOLD.sr + GFOLD.sen), 
               size = 2, 
               alpha = 0.7 ) + 
    scale_color_gradient2(low = "green", mid = "grey", high = "red", midpoint = 0) + 
    coord_cartesian(xlim = c(-2, 2), ylim = c(-2, 2)) + 
    theme_bw() +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank()) +
    labs(title = "Intersection of Up- and Down-regulated Genes")

sel <- (gfold.df$GFOLD.sr > gfold.cut & gfold.df$GFOLD.sen <= gfold.cut) | (gfold.df$GFOLD.sr < -gfold.cut & gfold.df$GFOLD.sen >= -gfold.cut)


g <- ggplot(gfold.df[sel, ], aes(GFOLD.sr, GFOLD.sen)) 
g + geom_point(aes(col = GFOLD.sr), 
               size = 2, 
               alpha = 0.7) + 
    scale_color_gradient2(low = "green", mid = "grey", high = "red", midpoint = 0) + 
    coord_cartesian(xlim = c(-2, 2), ylim = c(-2, 2)) + 
    theme_bw() +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank()) +
    labs(title = "Uniquely Up- and Down-regulated Genes in SR")

sel <- (gfold.df$GFOLD.sen > gfold.cut & gfold.df$GFOLD.sr <= gfold.cut) | (gfold.df$GFOLD.sen < -gfold.cut & gfold.df$GFOLD.sr >= -gfold.cut)


g <- ggplot(gfold.df[sel, ], aes(GFOLD.sr, GFOLD.sen)) 
g + geom_point(aes(col = GFOLD.sen), 
               size = 2, 
               alpha = 0.7) + 
    scale_color_gradient2(low = "green", mid = "grey", high = "red", midpoint = 0) + 
    coord_cartesian(xlim = c(-2, 2), ylim = c(-2, 2)) + 
    theme_bw() +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank()) +
    labs(title = "Uniquely Up- and Down-regulated Genes in SEN")

#dev.off()

```

```{r eval = FALSE}

g <- ggplot(gfold.df, aes(GFOLD.sr, GFOLD.sen))
g + stat_density2d(geom = "tile", n = 300, aes(fill = ..density..^0.2, alpha = 1), contour = FALSE) +
    geom_point(size = 0.5) +
    stat_density2d(geom = "tile", aes(fill = ..density..^0.2, alpha = ifelse(..density..^0.2 < 0.4, 0, 1)), contour = FALSE) + 
    scale_fill_gradientn(colours = colorRampPalette(c("white", blues9))(256)) + 
    coord_cartesian(xlim = c(-2, 2), ylim = c(-2, 2)) + 
    theme_bw() +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank())

```



## Pairwise distance between samples
```{r distSamples, message = FALSE, fig.height = 6}
library(gplots)
hmcol = colorRampPalette(c("green", "darkgreen", "black","darkred", "red"))(100)
dists = dist(t(gfold.df[, gfold.exprs]))
mat = as.matrix(dists)

hc <- hclust(dists)
par(mar = c(0, 5, 2, 1), cex = 1)
pdf("pairwise-dist-dendro.pdf", height = 5, width = 3)
plot(hc, ylab = "", xlab = "")
dev.off()

#pdf("pairwise-dist.pdf", height = 5, width = 5)
heatmap.2(mat, trace = "none", col = hmcol, margin = c(13, 13))
#dev.off()
```

# Expression Level of Individual Genes/Sequences

```{r IndividualGeneExpression, message = FALSE, fig.width = 10, fig.height = 4, warnings = FALSE}
library(bear)
#c6 <- c6.container
plotGenes <- function(x, data, dataset = "", groups){
    values <- data[data$hgnc_symbol == x, ]
    dataset <- ifelse(nchar(dataset) == 0, dataset, paste(" from ", dataset))
    if(nrow(values) == 0){
        print(paste("No data found for ", x))
    }
    else{
        cols <- grep("sr|sen|rej", colnames(values), perl = TRUE)
        
        values <- as.numeric(unlist(values)[cols])
        values.df <- data.frame(values, groups)
        sum.var <- summarySE(values.df, measurevar = "values", groupvars = "groups")
        
        g <- ggplot(sum.var, aes(x = factor(groups), y = values, fill = groups)) +
            geom_bar(stat = "identity") +
            geom_errorbar(aes(ymin = values - se, ymax = values + se),
                          size = 0.2,
                          width = 0.2) +
            scale_fill_manual(values = c6) +
            labs(title = paste(x, dataset), x = "", y = "Counts") + 
        theme_bw() +
        theme(panel.grid.major = element_blank(), 
              legend.position = "none",
              axis.text.x = element_text(angle = 45, hjust = 1))
        return(g)
    }
}



multiplot <- function(..., plotlist = NULL, file, cols = 1, layout = NULL) {
      require(grid)
    
      plots <- c(list(...), plotlist)
    
      numPlots = length(plots)
    
      if (is.null(layout)) {
        layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                        ncol = cols, nrow = ceiling(numPlots/cols))
      }
    
      if (numPlots == 1) {
        print(plots[[1]])
    
      } else {
        grid.newpage()
        pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
        for (i in 1:numPlots) {
          matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
    
          print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                          layout.pos.col = matchidx$col))
        }
      }
}

old.hk <- c("GAPDH", "RPL13A", "ACTB")

datasets <- list(counts.df, actb.df)
setnames <- c("Raw", "ACTB nm.")
names(datasets) <- setnames

plotMore <- function(datasets, setnames, genes){
    res <- lapply(genes, plotGenes, datasets, setnames, groups = groups)
    multiplot(plotlist = res, cols = 3 )
}

```

```{r message = FALSE, warning = FALSE, fig.height = 3, fig.wdith = 14}
Map(plotMore, datasets, setnames, rep(list(old.hk), length(datasets)))
```

# Functional Enrichment Analysis

## Gene Set Enrichment Analysis

### SR treated vs SR untreated

- [Gene sets enriched in SR treated Up][gs01]
- [Gene sets enriched in SR treated Down][gs02]

### SEN treated vs SEN untreated

- [Gene sets enriched in SEN treated Up][gs03]
- [Gene sets enriched in SEN treated Down][gs04]


## Gene Set Enrichment Analysis with Wilcoxon Test

- [SR treated vs SR untreated][gs05]

- [SEN treated vs SEN untreated][gs06]

### Pathway Graph

```{r eval = F}
# Down Pathway
tb <- read.csv("./pathway.graph.data/DownPathway-NodeAttributes.csv", header = T, stringsAsFactors = F)
pg <- strsplit(tb$SELF.REPLICATING, ", ")
names(pg) = tb$PATHWAY
pnames = tb$PATHWAY
res <- apply(combn(1:length(pg), 2), 2, function(x) {
    total = sum(pg[[x[1]]] %in% pg[[x[2]]])
    res = c(pnames[x[1]], pnames[x[2]], total)
})
df = as.data.frame(t(res), stringsAsFactors = F)
df1 = df[df$V3 > 0, ]
write.csv(df1, "./pathway.graph.data/DownPathway-EdgeAttributes.csv", quote = F, row.names = F)

# Up Pathway
## SR
tb <- read.csv("./pathway.graph.data/UpPathway-NodeAttributes-SR.csv", header = T, stringsAsFactors = F)
pg <- strsplit(tb$SELF.REPLICATING, ", ")
names(pg) = tb$PATHWAY
pnames = tb$PATHWAY
res <- apply(combn(1:length(pg), 2), 2, function(x) {
    total = sum(pg[[x[1]]] %in% pg[[x[2]]])
    res = c(pnames[x[1]], pnames[x[2]], total)
})
df = as.data.frame(t(res), stringsAsFactors = F)
df1 = df[df$V3 > 0, ]
write.csv(df1, "./pathway.graph.data/UpPathway-EdgeAttributes-SR.csv", quote = F, row.names = F)

## SEN
tb <- read.csv("./pathway.graph.data/UpPathway-NodeAttributes-SEN.csv", header = T, stringsAsFactors = F)
pg <- strsplit(tb$SELF.REPLICATING, ", ")
names(pg) = tb$PATHWAY
pnames = tb$PATHWAY
res <- apply(combn(1:length(pg), 2), 2, function(x) {
    total = sum(pg[[x[1]]] %in% pg[[x[2]]])
    res = c(pnames[x[1]], pnames[x[2]], total)
})
df = as.data.frame(t(res), stringsAsFactors = F)
df1 = df[df$V3 > 0, ]
write.csv(df1, "./pathway.graph.data/UpPathway-EdgeAttributes-SEN.csv", quote = F, row.names = F)

## INTERSECT
tb <- read.csv("./pathway.graph.data/UpPathwayIntersect-NodeAttributes.csv", header = T, stringsAsFactors = F)
pg <- strsplit(tb$SELF.REPLICATING, ", ")
names(pg) = tb$PATHWAY
pnames = tb$PATHWAY
res <- apply(combn(1:length(pg), 2), 2, function(x) {
    total = sum(pg[[x[1]]] %in% pg[[x[2]]])
    res = c(pnames[x[1]], pnames[x[2]], total)
})
df = as.data.frame(t(res), stringsAsFactors = F)
df1 = df[df$V3 > 0, ]
write.csv(df1, "./pathway.graph.data/UpPathwayIntersect-EdgeAttributes.csv", quote = F, row.names = F)
```

```{r geneSetsFromGSEA, eval = F}
library(reshape)
tb <- read.csv("c2.cp.biocarta.v4.0.symbols.gmt.csv", header = F, stringsAsFactors = F)
mylist <- strsplit(tb$V3, "\t")
names(mylist) = tb$V1

# Hypergeometric test
sr.up.only = as.character(gfold.df[gfold.up.sr.u, 2])
sr.up.only = sr.up.only[order(gfold.df[gfold.up.sr.u, "GFOLD.sr"], decreasing = T)]
sen.up.only = as.character(gfold.df[gfold.up.sen.u, 2])
sen.up.only = sen.up.only[order(gfold.df[gfold.up.sen.u, "GFOLD.sen"], decreasing = T)]
up.intersect = as.character(gfold.df[gfold.up.intersect, 2])
up.intersect = up.intersect[order(gfold.df[gfold.up.intersect, "GFOLD.sr"], decreasing = T)]

plist <- read.csv("./pathway.graph.data/plist-up-pathway-intersect.csv", header = T, stringsAsFactors = F)
# Remove white space
plist$PATHWAY = gsub(" ", "", plist$PATHWAY)

#paste(up.intersect[up.intersect %in% mylist$BIOCARTA_INTEGRIN_PATHWAY], collapse = ", ")
#sum(up.intersect %in% mylist$BIOCARTA_INTEGRIN_PATHWAY)

res = sapply(plist$PATHWAY, function(x) {
    res = paste(up.intersect[up.intersect %in% mylist[[x]]], collapse = ", ")
    res = c(res, sum(up.intersect %in% mylist[[x]]))
})



M = 2778 #M: number of up-reg gene sybols mapped by GSEA
knownGenes = 45956 #http://www.genecards.org/
N = knownGenes - M

#Example K=28, k=16
p.value = 1-phyper(16-1, M, knownGenes-M, 28)

#pvals <- lapply(mylist, 

# myHpg <- function(x, M=M, N=N, gene.set=sr.up.only){
#     K=length(x)
#     k=sum(gene.set %in% x)
#     p.value = 1-phyper(k-1, M, N, K)
# }
                
#mdf <- melt(mylist)
```


# Transcription Factor Enrichment Analysis

## ENCODEQT Results

- [ENCODEQT TF Enrichment Results][tf01]

# Downloadable Files

```{r}
write.csv(actb.df, "actb.normalized.counts.csv", quote = FALSE, row.names = FALSE)
```

[Differentially Expressed Protein-coding Genes Normalized by ACTB from GFOLD][f001]


# Meeting Notes

[20140523][002]

[20140606][003]

[20140828][004]

# Manuscript

Figure/Table | AI | PDF
----| ----| ----
Figure 1 | [AI][m0011] | [PDF][m0012]
Figure S1 | [AI][m0021] | [PDF][m0022]
Figure S2 | [AI][m0031] | [PDF][m0032]
Figure S3 | [AI][m0041] | [PDF][m0042]
Table S1 | [xlsx][m0051] | 


[img01]:Kazak_Overall_experimental_design.jpg

[f001]:./data/kazak.DEgenes.normbyACTB.xlsx
[f002]:./data/kazak.DEgenes.normbyACTB_Unique.xlsx

[001]:http://www.lifetechnologies.com/order/catalog/product/4456740
[002]:./slides/20140523.Overall_analytical_design.pdf
[003]:./slides/20140606.ERCC.pdf
[004]:./slides/20140828.Normalization_of_kazak_RNA-seq_data.pdf

[gs01]:./data/actb-gsea/sr_vs_sr_treated.Gsea.1412622136725/gsea_report_for_sr_treated_1412622136725.html
[gs02]:./data/actb-gsea/sr_vs_sr_treated.Gsea.1412622136725/gsea_report_for_sr_untreated_1412622136725.html

[gs03]:./data/actb-gsea/sen_vs_sen_treated.Gsea.1412622307486/gsea_report_for_sen_treated_1412622307486.html
[gs04]:./data/actb-gsea/sen_vs_sen_treated.Gsea.1412622307486/gsea_report_for_sen_untreated_1412622307486.html
[gs05]:./data/wilcoxon/SR-vs-SR-treated.html
[gs06]:./data/wilcoxon/SEN-vs-SEN-treated.html

[tf01]:tf-enrichment.html

[m0011]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.1.ai
[m0012]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.1.pdf
[m0021]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.S1.ai
[m0022]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.S1.pdf
[m0031]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.S2.ai
[m0032]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.S2.pdf
[m0041]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.S3.ai
[m0042]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Figure.S3.pdf
[m0051]:http://jordan.biology.gatech.edu/lulu/kazak-manuscript/Table.S1.xlsx
